### DISCUSSION TITLE
CondaPkg.toml in test directory

### DISCUSSION THREAD

**Discussion by icweaver** [cjdoris/PythonCall.jl#96] [Q&A] [answered]
Would it be possible to just have this file in the `test` directory for a package? It seems to only be found when it lives in the parent directory when running `pkg> test` on files that `use PythonCall` here. I was hoping to just have all Python related files isolated to the `test` directory if that seems reasonable

**Comment by cjdoris** [comment, marked as answer]
Do you mean that you want to declare some Conda dependencies which are only installed when running the tests? That would be reasonable.

I'm not sure exactly how Julia sets up the test environment. You can make the test folder be its own project (i.e. put a Project.toml there), maybe if you do that then it will get included into the load_path, and then CondaPkg will find dependencies in that folder. Maybe?

**Reply by icweaver** [reply]
Right! Thanks for the suggestion, it was something I was trying earlier and still seemed to be running into the same issue. Currently, CondaPkg.toml seems to be ignored when running `test` in package layouts like this for example:

```shell
MyPkg.jl/
├── Project.toml
├── src/
│   └── MyPkg.jl
└── test/
    ├── .CondaPkg [plan to just add this to .gitignore]
    ├── CondaPkg.toml
    ├── Project.toml
    └── runtests.jl
```

where `using PythonCall` lives in the contents of runtests.jl (or whatever subfile it `include`s). Apparently, `(MyPkg) pkg> test` does some ["magic"](https://github.com/JuliaLang/Pkg.jl/issues/1788#issuecomment-631976604) to combine the project and test specific envs together, so maybe that is interfering with how `PythonCall.jl` searches for `CondaPkg.toml` somehow?

**Reply by cjdoris** [reply]
If there's any magic involved then it's no surprise CondaPkg can't find CondaPkg.toml. The simplest solution is probably to use `CondaPkg.add` etc in `runtests.jl` to add the dependencies to the temporary test project. Or you can copy CondaPkg.toml from the test folder into the temporary test project (I guess CondaPkg could ship a function to do this for you.)

**Reply by icweaver** [reply]
Thank you for the workaround. After using CondaPkg.jl (and Julia's test harness) a little bit more now, I think I was mistaken when saying that PythonCall.jl could not find the `CondaPkg.toml` if it's stored in `test`. I think it's just not even checking there in the first place? `LOAD_PATH` and `Base.load_path()` seem to be automatically updated to the new tmp directory when running `(MyPkg) pkg> test`:

```
# @show LOAD_PATH, and @show Base.load_path() inside of test/runtests.jl

LOAD_PATH = ["@", "/tmp/jl_K84EE9"]
Base.load_path() = ["/tmp/jl_K84EE9/Project.toml"]
```

IIUC, if CondaPkg.jl does not see `CondaPkg.toml` here, then it will just create its own when making its own environment. Could an extra conditional be added to check for a test folder first:

```
/tmp/jl_K84EE9/test
``` 

before falling back to creating its own `CondaPkg.toml` in the root directory?

**Reply by cjdoris** [reply]
We'd only want to include `test/CondaPkg.toml` if we're actually running the tests. I wonder if it's possible to check if the tests are being run?

What's actually in that folder? Is it a copy of your project? Is the Project.toml modified in any way by Julia's testing machinery?

**Reply by icweaver** [reply]
> We'd only want to include test/CondaPkg.toml if we're actually running the tests

Do you mean that you'd like to be able to support multiple `CondaPkg.toml` files, one for the main project, then potentially one for the test directory if it exists? That'd be really cool!

> What's actually in that folder? Is it a copy of your project?

hmm, it looks like it's only showing `[".CondaPkg", "Manifest.toml", "Project.toml"]` in the tmp directory even though `CondaPkg.toml` is located in the original `test` folder as well. This is just from printing it out when `runtests.jl` runs though, so I guess things will be different during precompilation? I just uploaded a quick skeleton package that I am trying this on [here](https://github.com/icweaver/MyPkg_PythonCall.jl)

Just fair warning though that I am firmly outside of my level of familiarity with these sorts of things now 😅

**Reply by icweaver** [reply]
I thing having genuine per subdirectory `Project.toml`s in the future could be really cool, especially combined with `CondaPkg.jl`'s take on directory specific envs.

Until [#1233](https://github.com/JuliaLang/Pkg.jl/issues/1233) is merged, I think I am happy with just using the [pre-Julia v1.2](https://pkgdocs.julialang.org/v1/creating-packages/#Test-specific-dependencies-in-Julia-1.0-and-1.1) standard, as your package currently does! By the way, I stumbled on [this collapsed thread](https://user-images.githubusercontent.com/25312320/149644706-161d4656-042a-4b2d-a53d-865f48a90874.png) from [#499](https://github.com/JuliaDebug/JuliaInterpreter.jl/pull/499) that discusses some of the current pitfalls with the more "modern" approach of separate `Project.toml`s that I had never considered before. Here's hoping for a much cleaner approach in the future. Thanks again!

**Reply by cjdoris** [reply]
Thanks, those are some really useful links to understand the test mechanism. Even better, I think we concluded I don't need to do anything to CondaPkg. I'll add a note to the readme about this.

**Comment by icweaver** [comment]
ah, on second thought it doesn't look like Julia is moving the `test` directory to the tmp directory

**Reply by cjdoris** [reply]
I'm not sure there will be a reliable way to do this automatically, given the testing machinery is a bit magic. The best solution is probably to copy `YourProject/test/CondaPkg.toml` into the tmp directory at the top of `runtests.jl`.
